reader.go

```diff
@@ -122,76 +122,56 @@ func (r *Reader) Config() ReaderConfig {
 // messages from it.
 func (r *Reader) Close() error {
 	r.mutex.Lock()
-	if !r.closed {
-		close(r.done)
-	}
+	closed := r.closed
 	r.closed = true
-	r.dirty = false
-	r.cancel()
 	r.mutex.Unlock()
+
+	r.cancel()
+	r.join.Wait()
```

Comment:
When I want to write a test like this I generally wrap the WaitGroup in an interface with a local private struct in which you increment a counter and the waitgroup at the same time.  There are ways to do it which are clean.